File: backend/KtorApp/src/main/kotlin/com/example/utils/BackgroundProcessing.kt
package com.example.utils

import com.example.models.*
import org.bson.Document
import org.bson.types.ObjectId
import kotlinx.coroutines.*
import com.mongodb.client.model.Filters
import com.mongodb.client.model.Updates

object BackgroundProcessing {

    /**
     * Фоновая обработка загруженной книги.
     * @param idUser ID пользователя, загрузившего книгу.
     * @param filePath Путь к файлу книги на сервере.
     */
    /**
     * Фоновая обработка загруженной книги.
     * @param idUser ID пользователя, загрузившего книгу.
     * @param filePath Путь к файлу книги на сервере.
     * @param nameFile Название файла книги.
     */
    suspend fun processBookInBackground(idUser: String, filePath: String, nameFile: String) {
        withContext(Dispatchers.IO) {
            // Инициализация объектов для работы с книгой
            val bookProcessor = BookProcessor(filePath, nameFile)
            val chapters = bookProcessor.getChapters()
            var book = bookProcessor.getBook()

            // Создание записи о книге в базе данных
            val booksCollection = DatabaseFactory.getBooksCollection()
            booksCollection.insertOne(book.toDocument())

            val bookId = book.idBook

            // Обновление массива bookIds у пользователя
            val usersCollection = DatabaseFactory.getUsersCollection()
            usersCollection.updateOne(
                Filters.eq("_id", idUser),
                Updates.addToSet("bookIds", bookId)
            )

            // Разделение книги на главы и блоки
            val textBlocks = mutableListOf<Document>()
            for ((chapterIndex, chapter) in chapters.withIndex()) {
                val blocks = bookProcessor.divideChapterIntoBlocks(chapter)
                for (block in blocks) {
                    val idBlock = ObjectId().toString()
                    val textBlock = TextBlock(
                        idBlock = idBlock,
                        original = block,
                        numberChapter = chapterIndex + 1
                    )
                    textBlocks.add(textBlock.toDocument())

                    // Связывание блоков с книгой
                    booksCollection.updateOne(
                        Filters.eq("_id", bookId),
                        Updates.push("textBlockIds", idBlock)
                    )
                }
            }

            // Вставка текстовых блоков в базу данных
            val textBlocksCollection = DatabaseFactory.getTextBlocksCollection()
            if (textBlocks.isNotEmpty()) {
                textBlocksCollection.insertMany(textBlocks)
            }

            // Обновление статуса книги на "ready"
            booksCollection.updateOne(
                Filters.eq("_id", bookId),
                Updates.set("status", "ready")
            )
        }
    }

    /**
     * Фоновая обработка суммирования для отдельного текстового блока.
     * @param textBlock Текстовый блок для обработки.
     * @param mode Режим суммирования ("summarization" или "summarization_time").
     */
    suspend fun processSummarizationForBlock(textBlock: TextBlock, mode: String) {
        withContext(Dispatchers.IO) {
            val textBlocksCollection = DatabaseFactory.getTextBlocksCollection()
            val summarizedText = when (mode) {
                "summarization" -> summarizeText(textBlock.original)
                "summarization_time" -> summarizeTextWithTime(textBlock.original)
                else -> textBlock.original
            }

            val updateField = if (mode == "summarization") "summary" else "summaryTime"
            textBlocksCollection.updateOne(
                Filters.eq("_id", textBlock.idBlock),
                Updates.set(updateField, summarizedText)
            )
        }
    }

    /**
     * Фоновая обработка для режима суммирования.
     * @param book Книга, для которой выполняется суммирование.
     * @param mode Режим суммирования ("summarization" или "summarization_time").
     */
    suspend fun processSummarizationMode(book: Book, mode: String) {
        withContext(Dispatchers.IO) {
            val textBlocksCollection = DatabaseFactory.getTextBlocksCollection()

            // Получаем все текстовые блоки книги
            val blocksCursor = textBlocksCollection.find(Filters.`in`("_id", book.textBlockIds))
            val blocks = blocksCursor.map { it.toTextBlock() }.toList()

            // Обработка каждого блока
            for (block in blocks) {
                val summarizedText = when (mode) {
                    "summarization" -> summarizeText(block.original)
                    "summarization_time" -> summarizeTextWithTime(block.original)
                    else -> block.original
                }

                // Обновляем текстовый блок в базе данных
                val updateField = if (mode == "summarization") "summary" else "summaryTime"
                textBlocksCollection.updateOne(
                    Filters.eq("_id", block.idBlock),
                    Updates.set(updateField, summarizedText)
                )
            }

            // Обновление статуса книги
            val booksCollection = DatabaseFactory.getBooksCollection()
            booksCollection.updateOne(
                Filters.eq("_id", book.idBook),
                Updates.set("status", "summarized")
            )
        }
    }

    /**
     * Функция для суммирования текста.
     * @param text Оригинальный текст.
     * @return Суммированный текст.
     */
    fun summarizeText(text: String): String {
        // Реализуйте алгоритм суммирования текста
        // Здесь можно использовать библиотеку NLP или API
        return "Суммированный текст"
    }

    /**
     * Функция для суммирования текста с учетом времени чтения.
     * @param text Оригинальный текст.
     * @return Суммированный текст, оптимизированный по времени чтения.
     */
    fun summarizeTextWithTime(text: String): String {
        // Реализуйте алгоритм суммирования с учетом времени чтения
        return "Суммированный текст с учетом времени"
    }

    /**
     * Генерация вопросов и ответов по текстовому блоку.
     * @param text Текстовый блок.
     * @return Пара списков вопросов и ответов.
     */
    fun generateQuestionsForBlock(text: String): Pair<List<String>, List<String>> {
        // Реализуйте генерацию вопросов и ответов по тексту
        val questions = listOf("Вопрос 1 по блоку", "Вопрос 2 по блоку")
        val answers = listOf("Ответ 1", "Ответ 2")
        return Pair(questions, answers)
    }

    /**
     * Генерация теста по всей книге.
     * @param idBook ID книги.
     * @return Карта с вопросами и ответами.
     */
    suspend fun generateTestForBook(idBook: String): Map<String, Any> {
        val textBlocksCollection = DatabaseFactory.getTextBlocksCollection()

        // Получаем все текстовые блоки книги
        val blocksCursor = textBlocksCollection.find(Filters.eq("bookId", idBook))
        val blocks = blocksCursor.map { it.toTextBlock() }.toList()

        val questions = mutableListOf<String>()
        val answers = mutableListOf<String>()

        for (block in blocks) {
            val (blockQuestions, blockAnswers) = generateQuestionsForBlock(block.original)
            questions.addAll(blockQuestions)
            answers.addAll(blockAnswers)

            // Обновляем текстовый блок в базе данных
            textBlocksCollection.updateOne(
                Filters.eq("_id", block.idBlock),
                Updates.combine(
                    Updates.set("questions", blockQuestions),
                    Updates.set("rightAnswers", blockAnswers)
                )
            )
        }

        return mapOf("questions" to questions, "answers" to answers)
    }

    /**
     * Получение пересказа книги.
     * @param idBook ID книги.
     * @return Пересказ книги.
     */
    suspend fun getBookRetelling(idBook: String): String {
        val textBlocksCollection = DatabaseFactory.getTextBlocksCollection()

        // Получаем суммированные тексты всех блоков
        val blocksCursor = textBlocksCollection.find(Filters.eq("bookId", idBook))
        val summaries = blocksCursor.mapNotNull { it.getString("summary") }.toList()

        // Объединяем суммированные тексты в один пересказ
        return summaries.joinToString("\n")
    }

    /**
     * Получение рекомендаций похожих книг.
     * @param idBook ID книги.
     * @return Список рекомендованных книг.
     */
    suspend fun getSimilarBooks(idBook: String): List<Book> {
        // Реализуйте логику рекомендаций на основе жанра, автора или других параметров
        val booksCollection = DatabaseFactory.getBooksCollection()

        // Для примера выбираем случайные книги
        val randomBooksCursor = booksCollection.find().limit(5)
        return randomBooksCursor.map { it.toBook() }.toList()
    }
}

File: backend/utils/BackgroundProcessing.py
import asyncio
from pymongo import MongoClient
from bson import ObjectId
from pymongo.collection import ReturnDocument
from typing import List, Tuple

# Assuming the existence of these classes based on the Kotlin code
from models.Book import Book
from models.TextBlock import TextBlock
from utils.BookProcessor import BookProcessor
from utils.DatabaseFactory import DatabaseFactory

class BackgroundProcessing:

    @staticmethod
    async def process_book_in_background(idUser: str, filePath: str, nameFile: str, fileType: str):
        loop = asyncio.get_event_loop()
        await loop.run_in_executor(None, BackgroundProcessing._process_book_sync, idUser, filePath, nameFile, fileType)

    @staticmethod
    def _process_book_sync(idUser: str, filePath: str, nameFile: str, fileType: str):
        book_processor = BookProcessor(filePath, nameFile)
        chapters = book_processor.get_chapters(fileType)
        book = book_processor.get_book(fileType)

        books_collection = DatabaseFactory.get_books_collection()
        books_collection.insert_one(book.to_document())

        book_id = book.id_book

        users_collection = DatabaseFactory.get_users_collection()
        users_collection.update_one(
            {"_id": idUser},
            {"$addToSet": {"bookIds": book_id}}
        )

        text_blocks = []
        for chapter_index, chapter in enumerate(chapters):
            blocks = book_processor.divide_chapter_into_blocks(chapter)
            for block in blocks:
                text_block = TextBlock(
                    _id=str(ObjectId()),
                    original=block,
                    number_chapter=chapter_index + 1
                )
                text_blocks.append(text_block.to_document())

                books_collection.update_one(
                    {"_id": book_id},
                    {"$push": {"textBlockIds": text_block._id}}
                )

        text_blocks_collection = DatabaseFactory.get_text_blocks_collection()
        if text_blocks:
            text_blocks_collection.insert_many(text_blocks)

        books_collection.update_one(
            {"_id": book_id},
            {"$set": {"status": "ready"}}
        )

    @staticmethod
    async def process_summarization_for_block(text_block: TextBlock, mode: str):
        loop = asyncio.get_event_loop()
        await loop.run_in_executor(None, BackgroundProcessing._process_block_summarization_sync, text_block, mode)

    @staticmethod
    def _process_block_summarization_sync(text_block: TextBlock, mode: str):
        text_blocks_collection = DatabaseFactory.get_text_blocks_collection()
        if mode == "summarization":
            summarized_text = BackgroundProcessing.summarize_text(text_block.original)
        elif mode == "summarization_time":
            summarized_text = BackgroundProcessing.summarize_text_with_time(text_block.original)
        else:
            summarized_text = text_block.original

        update_field = "summary" if mode == "summarization" else "summaryTime"
        text_blocks_collection.update_one(
            {"_id": text_block._id},
            {"$set": {update_field: summarized_text}}
        )

    @staticmethod
    async def process_summarization_mode(book: Book, mode: str):
        loop = asyncio.get_event_loop()
        await loop.run_in_executor(None, BackgroundProcessing._process_summarization_mode_sync, book, mode)

    @staticmethod
    def _process_summarization_mode_sync(book: Book, mode: str):
        text_blocks_collection = DatabaseFactory.get_text_blocks_collection()
        blocks_cursor = text_blocks_collection.find({"_id": {"$in": book.textBlockIds}})
        blocks = [TextBlock.from_document(doc) for doc in blocks_cursor]

        for block in blocks:
            if mode == "summarization":
                summarized_text = BackgroundProcessing.summarize_text(block.original)
            elif mode == "summarization_time":
                summarized_text = BackgroundProcessing.summarize_text_with_time(block.original)
            else:
                summarized_text = block.original

            update_field = "summary" if mode == "summarization" else "summaryTime"
            text_blocks_collection.update_one(
                {"_id": block._id},
                {"$set": {update_field: summarized_text}}
            )

        books_collection = DatabaseFactory.get_books_collection()
        books_collection.update_one(
            {"_id": book.id_book},
            {"$set": {"status": "summarized"}}
        )

    @staticmethod
    def summarize_text(text: str) -> str:
        return "Суммированный текст"

    @staticmethod
    def summarize_text_with_time(text: str) -> str:
        return "Суммированный текст с учетом времени"

    @staticmethod
    def generate_questions_for_block(text: str) -> Tuple[List[str], List[str]]:
        questions = ["Вопрос 1 по блоку", "Вопрос 2 по блоку"]
        answers = ["Ответ 1", "Ответ 2"]
        return questions, answers

    @staticmethod
    async def generate_test_for_book(idBook: str) -> dict:
        text_blocks_collection = DatabaseFactory.get_text_blocks_collection()
        blocks_cursor = text_blocks_collection.find({"bookId": idBook})
        blocks = [TextBlock.from_document(doc) for doc in blocks_cursor]

        questions = []
        answers = []

        for block in blocks:
            block_questions, block_answers = BackgroundProcessing.generate_questions_for_block(block.original)
            questions.extend(block_questions)
            answers.extend(block_answers)

            text_blocks_collection.update_one(
                {"_id": block._id},
                {
                    "$set": {
                        "questions": block_questions,
                        "rightAnswers": block_answers
                    }
                }
            )

        return {"questions": questions, "answers": answers}

    @staticmethod
    async def get_book_retelling(idBook: str) -> str:
        text_blocks_collection = DatabaseFactory.get_text_blocks_collection()
        blocks_cursor = text_blocks_collection.find({"bookId": idBook})
        summaries = [doc.get("summary") for doc in blocks_cursor if doc.get("summary")]

        return "\n".join(summaries)

    @staticmethod
    async def get_similar_books(idBook: str) -> List[Book]:
        books_collection = DatabaseFactory.get_books_collection()
        random_books_cursor = books_collection.aggregate([{"$sample": {"size": 5}}])        
        return [Book.from_document(doc) for doc in random_books_cursor]